eth/consensus/clique/_utils.py

```diff
@@ -0,0 +1,154 @@
+import time
+from typing import (
+    Iterable,
+)
+
+from eth.abc import (
+    BlockHeaderAPI,
+)
+from eth.constants import (
+    EMPTY_UNCLE_HASH,
+    ZERO_ADDRESS,
+    ZERO_HASH32,
+)
+from eth_keys import keys
+from eth_keys.datatypes import PrivateKey
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    encode_hex,
+    to_list,
+    ValidationError,
+)
+
+from .constants import (
+    COMMON_ADDRESS_LENGTH,
+    NONCE_AUTH,
+    NONCE_DROP,
+    SIGNATURE_LENGTH,
+    VANITY_LENGTH,
+)
+from .datatypes import (
+    Snapshot,
+)
+
+
+@to_list
+def get_signers_at_checkpoint(header: BlockHeaderAPI) -> Iterable[Address]:
+    """
+    Read the list of signers from a checkpoint header.
+    """
+
+    signer_count = int(
+        (len(header.extra_data) - VANITY_LENGTH - SIGNATURE_LENGTH) / COMMON_ADDRESS_LENGTH
+    )
+
+    for i in range(signer_count):
+        yield Address(
+            header.extra_data[VANITY_LENGTH + i * COMMON_ADDRESS_LENGTH:][:COMMON_ADDRESS_LENGTH]
+        )
+
+
+def get_signature_hash(header: BlockHeaderAPI) -> Hash32:
+    """
+    Return the hash that is signed by the block producer. It is defined as the hash of
+    the ``header`` except that the last 65 bytes of the ``extra_data`` (the signature)
+    are removed before calculating the hash.
+    """
+    signature_header = header.copy(
+        extra_data=header.extra_data[:len(header.extra_data) - SIGNATURE_LENGTH]
+    )
+    return signature_header.hash
+
+
+def get_block_signer(header: BlockHeaderAPI) -> Address:
+    """
+    Return the address of the signer of the ``header``.
+    """
+
+    if len(header.extra_data) < SIGNATURE_LENGTH:
+        raise Exception("header.extra_data too short to contain signature")
+
+    signature_bytes = header.extra_data[-SIGNATURE_LENGTH:]
+
+    signature = keys.Signature(signature_bytes=signature_bytes)
+
+    signature_hash = get_signature_hash(header)
+    return signature.recover_public_key_from_msg_hash(signature_hash).to_canonical_address()
+
+
+def is_in_turn(signer: Address, snapshot: Snapshot, header: BlockHeaderAPI) -> bool:
+    """
+    Return ``True`` if the block was produced *in turn*, otherwise return ``False``.
+    """
+    sorted_signers = snapshot.get_sorted_signers()
+    offset = 0
+    while offset < len(sorted_signers) and sorted_signers[offset] != signer:
+        offset += 1
```

Comment:
Is this the same as

```try
    offset = sorted_signers.index(signer)
except ValueError:
    return False
else:
    return header.block_number % len(sorted_signers) == offset
```

Reply:
I suspect `Sequence.index` is pretty fast, but the `bisect` module also has tools for finding the right index in pre-sorted data structures.

Reply:
Yep, that should do it. I think I literally just took the Go code here and translated it one by one to Python which resulted in this less idiomatic code.